apps/emqx_retainer/src/emqx_retainer.erl

```diff
@@ -321,24 +321,31 @@ update_config(
     OldConf
 ) ->
     #{
-        backend := BackendCfg,
+        backend := #{
+            type := BackendType,
+            storage_type := StorageType
+        },
         msg_clear_interval := ClearInterval
     } = NewConf,
 
-    #{backend := OldBackendCfg} = OldConf,
-
-    StorageType = maps:get(type, BackendCfg),
-    OldStrorageType = maps:get(type, OldBackendCfg),
-    case OldStrorageType of
-        StorageType ->
+    #{
+        backend := #{
+            type := OldBackendType,
+            storage_type := OldStorageType
+        }
+    } = OldConf,
+    SameBackendType = BackendType =:= OldBackendType,
+    SameStorageType = StorageType =:= OldStorageType,
+    case SameBackendType andalso SameStorageType of
+        true ->
             State#{
                 clear_timer := check_timer(
                     ClearTimer,
                     ClearInterval,
                     clear_expired
                 )
             };
-        _ ->
+        false ->
```

Comment:
Looks like all the retained messages will be dropped, or?

Reply:
Apparently, they are not lost.  Tested manually and they persist after the config change, and after a VM restart (if `disc`).  I've added a test assertion for that: https://github.com/emqx/emqx/pull/9676/commits/3c21cb442e775a21e9b2ec35eaa61cd424129148